### PR TITLE
introduce a cache for subteamname->ID lookups for short bursts

### PR DESCRIPTION
**Pull request by maxtaco**
- you have to explicit opt into this cache, via new teamLoader parameter `AllowNameLookupBurstCache`

### CODE REVIEW

**Comment by maxtaco** [comment]
cc: @mmaxim

File: go/libkb/burst_cache.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	Contextified
+	locktab   LockTable
+	lru       *lru.Cache
+	cacheLife time.Duration
+	name      string
+}
+
+// BurstCacheKey is a key for a burst cache resource. Needs to implement the one
+// method --- String() --- used for turning the key into an LRU and LockTab key.
+type BurstCacheKey interface {
```
Comment (mlsteele) [inline go/libkb/burst_cache.go]: Idk if you want to collide with [Stringer](https://golang.org/pkg/fmt/#Stringer). Maybe `KeyString()`

File: go/libkb/burst_cache.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	String() string
+}
+
+// NewBurstCache makes a new burst cache with the given size and cacheLife.
+// name is for debugging purposes.
+func NewBurstCache(g *GlobalContext, sz int, dur time.Duration, name string) *BurstCache {
```
Comment (mlsteele) [inline go/libkb/burst_cache.go]: `name` is 2 long a var name. Mybe just `r` (bc it looks like the begnng of `n`)

File: go/libkb/burst_cache.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	lru, err := lru.New(sz)
+	if err != nil {
+		g.Log.Fatalf("Bad LRU Constructor: %s", err.Error())
```
Comment (mlsteele) [inline go/libkb/burst_cache.go]: panic. `lru` is nil at this point so it's going to panic later anyway.

File: go/libkb/interfaces.go [hunk 4]
```diff
@@ -608,6 +608,7 @@ type TeamLoader interface {
 	Delete(ctx context.Context, teamID keybase1.TeamID) error
 	// Untrusted hint of what a team's latest seqno is
 	HintLatestSeqno(ctx context.Context, id keybase1.TeamID, seqno keybase1.Seqno) error
+	ResolveNameToIDUntrusted(ctx context.Context, teamName keybase1.TeamName, public bool, allowCache bool) (id keybase1.TeamID, err error)
```
Comment (mlsteele) [inline go/libkb/interfaces.go]: This doesn't need to be advertised to the system, does it?

File: go/libkb/burst_cache.go (hunk 3 above)
Comment (maxtaco) [reply go/libkb/burst_cache.go]: doesn't fatalf panic?

File: go/libkb/burst_cache.go (hunk 2 above)
Comment (maxtaco) [reply go/libkb/burst_cache.go]: dude!

File: go/teams/loader.go [hunk 5]
```diff
@@ -108,6 +114,74 @@ func (l *TeamLoader) HintLatestSeqno(ctx context.Context, teamID keybase1.TeamID
...
+	}
+	if idPointer, ok := idVoidPointer.(*keybase1.TeamID); ok {
+		id = *idPointer
+	} else {
+		return keybase1.TeamID(""), errors.New("bad cast out of nameLookupBurstCache")
+	}
+	return id, nil
+}
+
+func resolveNameToIDUntrustedAPICall(ctx context.Context, g *libkb.GlobalContext, teamName keybase1.TeamName, public bool) (id keybase1.TeamID, err error) {
```
Comment (mlsteele) [inline go/teams/loader.go]: this and `LoaderContextG.resolveNameToIDUntrusted` are now duplicate

Comment (maxtaco) [reply go/teams/loader.go]: oh i didn't remove it?

File: go/libkb/burst_cache.go (hunk 3 above)
Comment (mlsteele) [reply go/libkb/burst_cache.go]: oh yeah. log.Fatalf says it logs then os.Exit, and log.Panicf logs then panic. Either's fine

File: go/libkb/interfaces.go (hunk 4 above)
Comment (mlsteele) [reply go/libkb/interfaces.go]: ah i see

File: go/libkb/burst_cache.go [hunk 6]
```diff
@@ -26,17 +26,19 @@ type BurstCacheKey interface {
 }
 
 // NewBurstCache makes a new burst cache with the given size and cacheLife.
-// name is for debugging purposes.
-func NewBurstCache(g *GlobalContext, sz int, dur time.Duration, name string) *BurstCache {
-	lru, err := lru.New(sz)
+// The cache will be at most cacheSize items long, and items will live in there for
+// at most cacheLife duration. For debug logging purposes, this cache will be known
+// as cacheName.
+func NewBurstCache(g *GlobalContext, cacheSize int, cacheLife time.Duration, cacheName string) *BurstCache {
```
Comment (mlsteele) [inline go/libkb/burst_cache.go]: fineyouwin

**Comment by maxtaco** [comment]
thx!

**Pull request merged**